### PR TITLE
[REQUEST FOR COMMENT] Add client assertions auth to the default oidc provider

### PR DESCRIPTION
**Pull request by prevostc**
## Description

Support client_assertion in generic OIDC provider

## Motivation and Context

Enterprise oidc often prefers client assertions instead of clientid/clientsecret auth on redeem and refresh
Also: #1834 

## How Has This Been Tested?

Untested for now

## Checklist:

- [ ] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [ ] I have created a feature (non-master) branch for my PR.
- [ ] I have written tests for my code changes.

### CODE REVIEW

File: providers/oidc.go [hunk 1]
```diff
@@ -18,12 +24,19 @@ type OIDCProvider struct {
 	*ProviderData
 
 	SkipNonce bool
+
+	UseAssertionAuthentication bool
```
Comment (tuunit) [inline providers/oidc.go:28]: I know this was part of my example in slack. I would prefer that we change this to an enum for better extensibility:

 ```suggestion
 type AuthenticationMethod int64
 
 const (
  // (default) https://oauth.net/2/client-authentication/
	ClientSecret AuthenticationMethod = iota
	// https://oauth.net/2/mtls/
	MutalTLS   
	// https://oauth.net/private-key-jwt/
	PrivateKeyJWT
)
 
	AuthenticationMethod AuthenticationMethod
```

Comment (tuunit) [reply providers/oidc.go:28]: MTLS is currently implemented as part of https://github.com/oauth2-proxy/oauth2-proxy/pull/1838

File: providers/oidc.go [hunk 2]
```diff
@@ -18,12 +24,19 @@ type OIDCProvider struct {
...
+	AssertionAuthJWTKey        *ecdsa.PrivateKey
+	AssertionAuthAlgorithm     string
+	AssertionAuthSigningMethod jwt.SigningMethod
+	AssertionAuthKeyId         string
+	AssertionAuthExpire        time.Duration
```
Comment (tuunit) [inline providers/oidc.go:33]: I'm thinking about putting this into it's own sub struct and maybe move it one level higher. As it is not just part of the oidc specification but oauth2 in general. So we should keep in mind. That other providers might want to support it in the future as well 🤔

Something like a generic config inside the provider struct.

```
type Provider struct {
  ...
  AuthenticationConfig 🤔
}
```

Comment (tuunit) [reply providers/oidc.go:33]: What do you think about that?

**Comment by prevostc** [comment]
Something like that?

**Comment by prevostc** [comment]
This version works locally for my provider, I would like to know if the general direction is the right one before spending more time on extended testing

**Comment by tuunit** [comment]
Hi @prevostc,

thanks for all the time you already put into the PR. I'll try to find some time soon to give it a proper review.

**Comment by jjlakis** [comment]
Hello, I tried to run this version with Azure OIDC and JWT token provided by workload identity webhook:
```
extraArgs:
  - --provider=oidc
  - --client-id=5b2583ce-9789-4fd2-970c-2cf711b648ed
  - --oidc-issuer-url=https://login.microsoftonline.com/66209a4a-80f3-4602-8126-2193115722f8/v2.0
  - --upstream=static://202
  - --reverse-proxy=true
  - --jwt-key-file=/var/run/secrets/azure/tokens/azure-identity-token
  - --authentication-method=private_key_jwt
```

Getting following error:
```
2024/01/07 09:40:39 http: panic serving 10.42.0.8:42916: runtime error: invalid memory address or nil pointer dereference
goroutine 227 [running]:
net/http.(*conn).serve.func1()
        /usr/local/go/src/net/http/server.go:1850 +0xbf
panic({0xdd1040, 0x16f58c0})
        /usr/local/go/src/runtime/panic.go:890 +0x262
github.com/oauth2-proxy/oauth2-proxy/v7/providers.(*OIDCProvider).makeAssertionToken(0xc0003ec070)
        /go/src/github.com/oauth2-proxy/oauth2-proxy/providers/oidc.go:114 +0x91
github.com/oauth2-proxy/oauth2-proxy/v7/providers.(*OIDCProvider).RedeemAssertion(0xc0003ec070, {0x10c3f78, 0xc0001a3350}, {0xc00023b140, 0x2a}, {0xc00017271a, 0x5c1}, {0x0, 0x0})
        /go/src/github.com/oauth2-proxy/oauth2-proxy/providers/oidc.go:143 +0xa7
github.com/oauth2-proxy/oauth2-proxy/v7/providers.(*OIDCProvider).Redeem(0xc000220000?, {0x10c3f78?, 0xc0001a3350?}, {0xc00023b140?, 0xc000193450?}, {0xc00017271a?, 0xc000263ea0?}, {0x0?, 0x9bb18a?})
        /go/src/github.com/oauth2-proxy/oauth2-proxy/providers/oidc.go:76 +0x53
main.(*OAuthProxy).redeemCode(0xc000220000, 0xc0001ac400, {0x0, 0x0})
        /go/src/github.com/oauth2-proxy/oauth2-proxy/oauthproxy.go:901 +0x143
main.(*OAuthProxy).OAuthCallback(0xc000220000, {0x7f87d36f9a38, 0xc0001a33b0}, 0xc0001ac400)
        /go/src/github.com/oauth2-proxy/oauth2-proxy/oauthproxy.go:834 +0x372
net/http.HandlerFunc.ServeHTTP(0x7f87d36f9a38?, {0x7f87d36f9a38?, 0xc0001a33b0?}, 0x10c0b40?)
        /usr/local/go/src/net/http/server.go:2109 +0x2f
main.prepareNoCacheMiddleware.func1({0x7f87d36f9a38, 0xc0001a33b0}, 0x4c5357?)
        /go/src/github.com/oauth2-proxy/oauth2-proxy/oauthproxy.go:1014 +0x5c
net/http.HandlerFunc.ServeHTTP(0x8?, {0x7f87d36f9a38?, 0xc0001a33b0?}, 0x0?)
        /usr/local/go/src/net/http/server.go:2109 +0x2f
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerDuration.func2({0x7f87d36f9a38, 0xc0001a33b0}, 0xc0001ac400)
        /go/pkg/mod/github.com/prometheus/client_golang@v1.15.1/prometheus/promhttp/instrument_server.go:109 +0xc7
net/http.HandlerFunc.ServeHTTP(0xaec53e?, {0x7f87d36f9a38?, 0xc0001a33b0?}, 0xc0001a3380?)
        /usr/local/go/src/net/http/server.go:2109 +0x2f
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerInFlight.func1({0x7f87d36f9a38, 0xc0001a33b0}, 0x10c2a00?)
        /go/pkg/mod/github.com/prometheus/client_golang@v1.15.1/prometheus/promhttp/instrument_server.go:60 +0xd4
net/http.HandlerFunc.ServeHTTP(0x10c2a20?, {0x7f87d36f9a38?, 0xc0001a33b0?}, 0x0?)
        /usr/local/go/src/net/http/server.go:2109 +0x2f
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerCounter.func1({0x10c2a20?, 0xc0000435a0?}, 0xc0001ac400)
        /go/pkg/mod/github.com/prometheus/client_golang@v1.15.1/prometheus/promhttp/instrument_server.go:147 +0xc5
net/http.HandlerFunc.ServeHTTP(0xc000227320?, {0x10c2a20?, 0xc0000435a0?}, 0x0?)
        /usr/local/go/src/net/http/server.go:2109 +0x2f
github.com/oauth2-proxy/oauth2-proxy/v7/pkg/middleware.readynessCheck.func1({0x10c2a20?, 0xc0000435a0?}, 0xc0001ac400?)
        /go/src/github.com/oauth2-proxy/oauth2-proxy/pkg/middleware/readynesscheck.go:38 +0x238
net/http.HandlerFunc.ServeHTTP(0x10?, {0x10c2a20?, 0xc0000435a0?}, 0x10?)
        /usr/local/go/src/net/http/server.go:2109 +0x2f
github.com/oauth2-proxy/oauth2-proxy/v7/pkg/middleware.healthCheck.func1({0x10c2a20, 0xc0000435a0}, 0x4c5301?)
        /go/src/github.com/oauth2-proxy/oauth2-proxy/pkg/middleware/healthcheck.go:40 +0xbd
net/http.HandlerFunc.ServeHTTP(0x0?, {0x10c2a20?, 0xc0000435a0?}, 0x0?)
        /usr/local/go/src/net/http/server.go:2109 +0x2f
github.com/oauth2-proxy/oauth2-proxy/v7/pkg/middleware.requestLogger.func1({0x10c33b0?, 0xc00015a7e0}, 0xc0001ac400)
        /go/src/github.com/oauth2-proxy/oauth2-proxy/pkg/middleware/request_logger.go:27 +0x127
net/http.HandlerFunc.ServeHTTP(0xc0001ac300?, {0x10c33b0?, 0xc00015a7e0?}, 0x10b7f01?)
        /usr/local/go/src/net/http/server.go:2109 +0x2f
github.com/oauth2-proxy/oauth2-proxy/v7/pkg/middleware.NewScope.func1.1({0x10c33b0, 0xc00015a7e0}, 0xc0001a3320?)
        /go/src/github.com/oauth2-proxy/oauth2-proxy/pkg/middleware/scope.go:19 +0xcc
net/http.HandlerFunc.ServeHTTP(0xc0001ac200?, {0x10c33b0?, 0xc00015a7e0?}, 0xc000417a58?)
        /usr/local/go/src/net/http/server.go:2109 +0x2f
github.com/gorilla/mux.(*Router).ServeHTTP(0xc0002563c0, {0x10c33b0, 0xc00015a7e0}, 0xc0001ac100)
        /go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210 +0x1cf
main.(*OAuthProxy).ServeHTTP(0xc000172715?, {0x10c33b0?, 0xc00015a7e0?}, 0x4681ae?)
        /go/src/github.com/oauth2-proxy/oauth2-proxy/oauthproxy.go:533 +0x2a
net/http.serverHandler.ServeHTTP({0xc0001a2c30?}, {0x10c33b0, 0xc00015a7e0}, 0xc0001ac100)
        /usr/local/go/src/net/http/server.go:2947 +0x30c
net/http.(*conn).serve(0xc00023f900, {0x10c3f78, 0xc0001820f0})
        /usr/local/go/src/net/http/server.go:1991 +0x607
created by net/http.(*Server).Serve
        /usr/local/go/src/net/http/server.go:3102 +0x4db
[2024/01/07 09:40:39] [director.go:85] Invalid redirect generated from X-Forwarded-* headers: https://podinfo.lakis.tech/favicon.ico
[2024/01/07 09:40:39] [oauthproxy.go:972] No valid authentication in request. Initiating login.
[2024/01/07 09:40:39] [validator.go:60] Rejecting invalid redirect "https://podinfo.lakis.tech/favicon.ico": domain / port not in whitelist
10.42.0.1 - ba2c24ce-2c97-4811-8d39-6be355afe07d - - [2024/01/07 09:40:39] podinfo.lakis.tech GET - "/" HTTP/1.1 "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36" 403 8498 0.000
```

**Comment by prevostc** [comment]
Hi @jjlakis thanks for testing that PR.

Unfortunately I won't be updating this pull request as my company decided to roll it's own version of auth proxy in a programming language we're familiar with so we get full control over the delivery cycle and how fast we can push changes like this one.

For that reason I'll be closing this pull request but feel free to use the code.

**Comment by tuunit** [comment]
Hi @prevostc,

thank you for the time you invested into this PR. I'll take over further changes.

**Pull request closed without merging**